File: apps/workbench/backend/api/v1/evidence.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from fastapi import APIRouter, Header, HTTPException, Request, Response, status
from pydantic import BaseModel, ConfigDict, Field

from sec_agent.workbench.api_contracts import request_trace_id

from ...application.case_service import CasePrincipal
from ...application.evidence_service import (
    CompileEvidenceFixtureDraft,
    EvidenceReviewDraft,
    EvidenceService,
    EvidenceServiceError,
)


class CompileEvidenceFixtureCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_workspace_version: int = Field(ge=0)
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class EvidenceReviewCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_workspace_version: int = Field(ge=1)
    reason: str = Field(min_length=1)
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class EvidenceCandidateView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    candidate_id: str
    display_state: Literal["candidate", "context_only", "rejected"]
    candidate_kind: str
    title: str
    source_name: str
    source_type: str
    published_at: str
    citation: str
    excerpt: str
    authority_label: str
    source_authority: str
    source_role: str
    source_authority_rank: int
    source_policy_ref: str
    route_id: str
    document_id: str
    document_version: str
    entity_ref: str
    period_ref: str
    section_or_table_ref: str
    content_ref: str
    applicability_boundary: str
    promotion_boundary: Literal["not_in_Point03_VT1"]
    review_reason: str | None = None


class EvidenceSlotWorkbenchView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    evidence_slot_id: str
    evidence_role: str
    cell_id: str
    decision_question: str
    owner: str
    required: bool
    display_state: Literal["candidate", "typed_gap", "repair_requested"]
    request_id: str
    request_digest: str
    tool_plan_id: str
    tool_plan_status: str
    bundle_id: str
    bundle_status: str
    exhaustion_status: str
    typed_gap_codes: list[str]
    candidates: list[EvidenceCandidateView]


class EvidenceCountsView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    slot_count: int
    total_candidate_count: int
    candidate_count: int
    context_only_count: int
    rejected_count: int
    typed_gap_count: int
    repair_requested_count: int
    repair_completed_count: int
    review_action_count: int


class EvidenceReviewActionView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    review_action_id: str
    action_type: Literal["reject_candidate", "request_repair"]
    evidence_slot_id: str
    candidate_id: str | None
    reason: str
    actor_ref: str
    recorded_at: str
    workspace_version_after: int


class EvidenceRepairOutcomeView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    repair_outcome_id: str
    evidence_slot_id: str
    request_review_action_id: str
    attempt_no: int
    attempt_state: str
    route_id: str
    candidate_id: str
    completed_at: str
    external_call_count: int
    tool_invocation_count: int
    boundary: str


class EvidenceWorkbenchView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    workspace_id: str
    workspace_version: int
    projection_version_id: str
    status: str
    fixture_mode: Literal["fixture_shadow_internal_only"]
    decision_surface_contract_version_id: str
    planning_checkpoint_version_id: str
    work_unit_id: str
    counts: EvidenceCountsView
    slots: list[EvidenceSlotWorkbenchView]
    review_actions: list[EvidenceReviewActionView]
    repair_outcomes: list[EvidenceRepairOutcomeView]
    available_actions: list[Literal["reject_candidate", "request_repair", "execute_repair"]]
    hard_boundaries: dict[str, int | str]


def build_evidence_router(service: EvidenceService) -> APIRouter:
    router = APIRouter(tags=["point03-evidence"])

    @router.get(
        "/cases/{case_id}/evidence",
        operation_id="getEvidenceWorkbench",
        response_model=EvidenceWorkbenchView,
    )
    def get_evidence(
        case_id: str,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.get_workbench(
                case_id, _principal(tenant_id, project_id, actor_id, permissions)
            )
        except EvidenceServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _etag(projection)
        return projection

    @router.post(
        "/cases/{case_id}/evidence/compile",
        operation_id="compileEvidenceFixture",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=EvidenceWorkbenchView,
    )
    def compile_evidence(
        case_id: str,
        command: CompileEvidenceFixtureCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.compile_fixture(
                case_id,
                CompileEvidenceFixtureDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except EvidenceServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _etag(projection)
        return projection

    @router.post(
        "/cases/{case_id}/evidence/candidates/{candidate_id}/reject",
        operation_id="rejectEvidenceCandidate",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=EvidenceWorkbenchView,
    )
    def reject_candidate(
        case_id: str,
        candidate_id: str,
        command: EvidenceReviewCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.reject_candidate(
                case_id,
                candidate_id,
                EvidenceReviewDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except EvidenceServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _etag(projection)
        return projection

    @router.post(
        "/cases/{case_id}/evidence/slots/{evidence_slot_id}/request-repair",
        operation_id="requestEvidenceRepair",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=EvidenceWorkbenchView,
    )
    def request_repair(
        case_id: str,
        evidence_slot_id: str,
        command: EvidenceReviewCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.request_repair(
                case_id,
                evidence_slot_id,
                EvidenceReviewDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except EvidenceServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _etag(projection)
        return projection

    return router


def _principal(
    tenant_id: str | None,
    project_id: str | None,
    actor_id: str | None,
    permissions: str | None,
) -> CasePrincipal:
    return CasePrincipal(
        tenant_id=(tenant_id or "").strip(),
        project_id=(project_id or "").strip(),
        actor_id=(actor_id or "").strip(),
        permissions=frozenset(item.strip() for item in (permissions or "").split(",") if item.strip()),
    )


def _raise_service_error(error: EvidenceServiceError) -> None:
    raise HTTPException(status_code=error.status_code, detail=error.detail) from error


def _etag(projection: dict[str, Any]) -> str:
    return f'"evidence-workspace={projection["workspace_version"]}"'

File: apps/workbench/frontend/vite/src/api/evidence.ts
import { CASES_PATH, FixtureCaseContext } from "./cases";

export type EvidenceCandidateState = "candidate" | "context_only" | "rejected";
export type EvidenceGapState = "typed_gap" | "repair_requested";
export type EvidenceWorkbenchStatus = "not_prepared" | "prepared";

export type CompileEvidenceFixtureCommand = {
  expected_workspace_version: number;
  actor_ref: string;
  idempotency_key: string;
};

export type RejectEvidenceCandidateCommand = {
  expected_workspace_version: number;
  reason: string;
  actor_ref: string;
  idempotency_key: string;
};

export type RequestEvidenceRepairCommand = {
  expected_workspace_version: number;
  reason: string;
  actor_ref: string;
  idempotency_key: string;
};

export type ExecuteEvidenceRepairCommand = {
  expected_workspace_version: number;
  actor_ref: string;
  idempotency_key: string;
};

export type EvidenceWorkbenchSummary = {
  candidate_count: number;
  context_only_count: number;
  rejected_count: number;
  gap_count: number;
  repair_requested_count: number;
  repair_completed_count: number;
};

export type EvidenceRepairOutcomeView = {
  repair_outcome_id: string;
  evidence_slot_id: string;
  status: string;
  candidate_id: string;
  route_id: string;
  outcome_boundary: string;
  completed_at: string;
};

export type EvidenceCandidateView = {
  candidate_id: string;
  evidence_slot_id: string;
  state: EvidenceCandidateState;
  candidate_kind: string;
  title: string;
  excerpt: string;
  source_name: string;
  source_type: string;
  published_at: string;
  citation: string;
  authority_label: string;
  source_authority_rank: number;
  applicability_boundary: string;
  document_id: string;
  document_version: string;
  section_or_table_ref: string;
  source_policy_ref: string;
  route_id: string;
  source_role: string;
  entity_ref: string;
  period_ref: string;
  rejection_reason?: string | null;
  reviewed_at?: string | null;
};

export type EvidenceGapView = {
  gap_id: string;
  evidence_slot_id: string;
  state: EvidenceGapState;
  gap_code: string;
  title: string;
  detail: string;
  stop_rule: string;
  repair_reason?: string | null;
  requested_at?: string | null;
};

export type EvidenceCellSectionView = {
  cell_id: string;
  decision_question: string;
  owner: string;
  materiality: string;
  evidence_slot_id: string;
  evidence_role: string;
  request_id: string;
  bundle_status: string;
  candidates: EvidenceCandidateView[];
  typed_gap?: EvidenceGapView | null;
};

export type EvidenceWorkbenchView = {
  case_id: string;
  workspace_version: number;
  case_version: number;
  decision_surface_contract_version: number;
  checkpoint_version: number;
  work_unit_id: string;
  work_unit_version: number;
  status: EvidenceWorkbenchStatus;
  prepared_at?: string | null;
  summary: EvidenceWorkbenchSummary;
  cells: EvidenceCellSectionView[];
  repair_outcomes: EvidenceRepairOutcomeView[];
};

export type LocalResearchCandidateView = {
  candidate_id: string;
  retrieval_lane: "object_bm25" | "gold_fact_sql" | "research_graph";
  rank: number;
  score: number | null;
  ticker: string;
  title: string;
  excerpt: string;
  source_name: string;
  source_type: string;
  published_at: string;
  citation_url: string;
  citation_span: string;
  evidence_ref: string;
  authority_mode: string;
  claim_boundary: string;
  exact_value_authority: boolean;
  numeric_eligible: boolean;
  metric_family?: string;
  value?: string;
  unit?: string;
  period?: string;
  writer_citable: false;
  promotion_status: "candidate_not_promoted";
};

export type LocalResearchCellView = {
  cell_key: string;
  evidence_role: string;
  decision_question: string;
  retrieval_lane: LocalResearchCandidateView["retrieval_lane"];
  status: "candidate_ready" | "typed_gap";
  typed_gap: string | null;
  candidates: LocalResearchCandidateView[];
};

export type LocalResearchPreviewView = {
  preview_digest: string;
  case_id: string;
  case_version: number;
  query: string;
  as_of: string;
  research_mode: "bounded_local_read_only";
  status: "candidate_preview_ready";
  selected_cell_count: number;
  candidate_count: number;
  cells: LocalResearchCellView[];
  source_inventory: Array<{
    source_id: string;
    schema_version: string;
    record_count: number;
    snapshot_digest: string;
  }>;
  execution_counts: Record<string, number>;
  boundary: string;
};

export type LocalAnalysisFactView = {
  candidate_id: string;
  metric_family: string;
  label: string;
  value: string;
  unit: string;
  period: string;
  source_ref: string;
  exact_value_authority: boolean;
};

export type LocalAnalysisDerivedMetricView = {
  metric: string;
  label: string;
  value: string;
  unit: string;
  formula: string;
  input_candidate_ids: string[];
};

export type LocalAnalysisRepairView = {
  repair_id: string;
  evidence_role: string;
  decision: string;
  reason: string;
  candidate_refs: string[];
  remaining_gap: string;
  external_execution: false;
  promotion_authorized: false;
};

export type LocalAnalysisJudgmentView = {
  judgment_id: string;
  cell_key: string;
  evidence_role: string;
  decision_question: string;
  confidence: string;
  judgment_zh_cn: string;
  judgment_en: string;
  evidence_refs: string[];
  numeric_refs: string[];
  repair_ref: string;
  counter_thesis_zh_cn: string;
  what_would_change_en: string;
  remaining_gaps: string[];
  status: string;
};

export type LocalAnalysisPreviewView = {
  analysis_digest: string;
  case_id: string;
  case_version: number;
  source_preview_digest: string;
  analysis_mode: "bounded_local_deterministic_preview";
  status: "internal_analysis_preview_ready";
  numeric: {
    status: string;
    facts: LocalAnalysisFactView[];
    derived_metrics: LocalAnalysisDerivedMetricView[];
    typed_gaps: string[];
    writer_citable: false;
  };
  repairs: LocalAnalysisRepairView[];
  judgments: LocalAnalysisJudgmentView[];
  workpaper: {
    content_digest: string;
    status: string;
    senior_r2_status: string;
    sections: Array<Record<string, unknown>>;
  };
  writer: {
    content_digest: string;
    mode: "deterministic_no_source_internal_composer";
    status: string;
    title_zh_cn: string;
    title_en: string;
    sections: Array<{
      section_id: string;
      evidence_role: string;
      judgment_ref: string;
      heading_zh_cn: string;
      content_zh_cn: string;
      content_en: string;
    }>;
    input_workpaper_digest: string;
    source_preview_digest: string;
    source_access_calls: 0;
    model_calls: 0;
    release_admitted: false;
  };
  execution_counts: Record<string, number>;
  hard_boundaries: Record<string, number>;
  boundary: string;
};

const localResearchPreviewInFlight = new Map<string, Promise<LocalResearchPreviewView>>();
const localAnalysisPreviewInFlight = new Map<string, Promise<LocalAnalysisPreviewView>>();

type EvidenceWireCandidate = {
  candidate_id: string;
  display_state: EvidenceCandidateState;
  candidate_kind: string;
  title: string;
  excerpt: string;
  source_name: string;
  source_type: string;
  published_at: string;
  citation: string;
  authority_label: string;
  source_authority_rank: number;
  applicability_boundary: string;
  document_id: string;
  document_version: string;
  section_or_table_ref: string;
  source_policy_ref: string;
  route_id: string;
  source_role: string;
  entity_ref: string;
  period_ref: string;
  review_reason?: string | null;
};

type EvidenceWireSlot = {
  cell_id: string;
  decision_question: string;
  owner: string;
  required: boolean;
  evidence_slot_id: string;
  evidence_role: string;
  request_id: string;
  bundle_id: string;
  bundle_status: string;
  display_state: "candidate" | EvidenceGapState;
  typed_gap_codes: string[];
  candidates: EvidenceWireCandidate[];
};

type EvidenceWireReviewAction = {
  action_type: "reject_candidate" | "request_repair";
  evidence_slot_id: string;
  candidate_id?: string | null;
  reason: string;
  recorded_at: string;
};

type EvidenceWireRepairOutcome = {
  repair_outcome_id: string;
  evidence_slot_id: string;
  request_review_action_id: string;
  attempt_no: number;
  attempt_state: string;
  route_id: string;
  candidate_id: string;
  completed_at: string;
  external_call_count: number;
  tool_invocation_count: number;
  boundary: string;
};

type EvidenceWireView = {
  case_id: string;
  workspace_version: number;
  status: string;
  work_unit_id: string;
  counts: {
    candidate_count: number;
    context_only_count: number;
    rejected_count: number;
    typed_gap_count: number;
    repair_requested_count: number;
    repair_completed_count: number;
  };
  slots: EvidenceWireSlot[];
  review_actions: EvidenceWireReviewAction[];
  repair_outcomes: EvidenceWireRepairOutcome[];
};

type ErrorEnvelope = {
  error?: {
    error_code?: string;
    message?: string;
    status_code?: number;
    trace_id?: string;
  };
  code?: string;
  message?: string;
  correlation_id?: string;
  detail?: unknown;
};

export class EvidenceApiError extends Error {
  readonly code: string;
  readonly statusCode: number;
  readonly traceId: string | undefined;
  readonly detail: unknown;

  constructor(input: { code: string; message: string; statusCode: number; traceId?: string; detail?: unknown }) {
    super(input.message);
    this.name = "EvidenceApiError";
    this.code = input.code;
    this.statusCode = input.statusCode;
    this.traceId = input.traceId;
    this.detail = input.detail;
  }
}

export const DEFAULT_FIXTURE_EVIDENCE_CONTEXT: FixtureCaseContext = {
  tenantId: "fixture_internal",
  projectId: "workbench_internal",
  actorId: "analyst_internal",
  permissions: [
    "case:read",
    "planning:read",
    "execution:read",
    "evidence:read",
    "evidence:write",
    "evidence:review",
    "evidence:repair",
    "numeric:read",
  ],
};

export class EvidenceApiClient {
  constructor(private readonly context: FixtureCaseContext = DEFAULT_FIXTURE_EVIDENCE_CONTEXT) {}

  get actorRef(): string {
    return this.context.actorId;
  }

  async getEvidenceWorkbench(caseId: string): Promise<EvidenceWorkbenchView> {
    return normalizeEvidenceWorkbench(await this.request<EvidenceWireView>(evidencePath(caseId)));
  }

  async getLocalResearchPreview(caseId: string): Promise<LocalResearchPreviewView> {
    return shareInFlight(
      localResearchPreviewInFlight,
      `${this.context.tenantId}:${this.context.projectId}:${this.context.actorId}:${caseId}`,
      () => this.request<LocalResearchPreviewView>(
        `${CASES_PATH}/${encodeURIComponent(caseId)}/local-research-preview`,
      ),
    );
  }

  async getLocalAnalysisPreview(caseId: string): Promise<LocalAnalysisPreviewView> {
    return shareInFlight(
      localAnalysisPreviewInFlight,
      `${this.context.tenantId}:${this.context.projectId}:${this.context.actorId}:${caseId}`,
      () => this.request<LocalAnalysisPreviewView>(
        `${CASES_PATH}/${encodeURIComponent(caseId)}/local-analysis-preview`,
      ),
    );
  }

  async compileEvidenceFixture(caseId: string, command: CompileEvidenceFixtureCommand): Promise<EvidenceWorkbenchView> {
    return normalizeEvidenceWorkbench(
      await this.request<EvidenceWireView>(`${evidencePath(caseId)}/compile`, mutationInit(command)),
    );
  }

  async rejectEvidenceCandidate(
    caseId: string,
    candidateId: string,
    command: RejectEvidenceCandidateCommand,
  ): Promise<EvidenceWorkbenchView> {
    return normalizeEvidenceWorkbench(
      await this.request<EvidenceWireView>(
        `${evidencePath(caseId)}/candidates/${encodeURIComponent(candidateId)}/reject`,
        mutationInit(command),
      ),
    );
  }

  async requestEvidenceRepair(
    caseId: string,
    evidenceSlotId: string,
    command: RequestEvidenceRepairCommand,
  ): Promise<EvidenceWorkbenchView> {
    return normalizeEvidenceWorkbench(
      await this.request<EvidenceWireView>(
        `${evidencePath(caseId)}/slots/${encodeURIComponent(evidenceSlotId)}/request-repair`,
        mutationInit(command),
      ),
    );
  }

  async executeEvidenceRepair(
    caseId: string,
    evidenceSlotId: string,
    command: ExecuteEvidenceRepairCommand,
  ): Promise<EvidenceWorkbenchView> {
    return normalizeEvidenceWorkbench(
      await this.request<EvidenceWireView>(
        `${evidencePath(caseId)}/slots/${encodeURIComponent(evidenceSlotId)}/execute-repair`,
        mutationInit(command),
      ),
    );
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    headers.set("X-Fin-Case-Tenant", this.context.tenantId);
    headers.set("X-Fin-Case-Project", this.context.projectId);
    headers.set("X-Fin-Case-Actor", this.context.actorId);
    headers.set("X-Fin-Case-Permissions", this.context.permissions.join(","));

    let response: Response;
    try {
      response = await fetch(path, { ...init, headers });
    } catch (error) {
      throw new EvidenceApiError({
        code: "network_unavailable",
        message: "The fixture Evidence API is unavailable.",
        statusCode: 0,
        detail: error,
      });
    }

    const text = await response.text();
    const payload = text ? parsePayload(text) : {};
    if (!response.ok) throw toEvidenceApiError(response.status, payload);
    return payload as T;
  }
}

function shareInFlight<T>(registry: Map<string, Promise<T>>, key: string, request: () => Promise<T>): Promise<T> {
  const current = registry.get(key);
  if (current) return current;
  const pending = request();
  registry.set(key, pending);
  const release = () => {
    if (registry.get(key) === pending) registry.delete(key);
  };
  void pending.then(release, release);
  return pending;
}

function normalizeEvidenceWorkbench(wire: EvidenceWireView): EvidenceWorkbenchView {
  const repairBySlot = new Map(
    wire.review_actions
      .filter((action) => action.action_type === "request_repair")
      .map((action) => [action.evidence_slot_id, action]),
  );
  return {
    case_id: wire.case_id,
    workspace_version: wire.workspace_version,
    case_version: 0,
    decision_surface_contract_version: 0,
    checkpoint_version: 0,
    work_unit_id: wire.work_unit_id,
    work_unit_version: 1,
    status: "prepared",
    summary: {
      candidate_count: wire.counts.candidate_count,
      context_only_count: wire.counts.context_only_count,
      rejected_count: wire.counts.rejected_count,
      gap_count: wire.counts.typed_gap_count,
      repair_requested_count: wire.counts.repair_requested_count,
      repair_completed_count: wire.counts.repair_completed_count,
    },
    repair_outcomes: wire.repair_outcomes.map((outcome) => ({
      repair_outcome_id: outcome.repair_outcome_id,
      evidence_slot_id: outcome.evidence_slot_id,
      status: outcome.attempt_state,
      candidate_id: outcome.candidate_id,
      route_id: outcome.route_id,
      outcome_boundary: outcome.boundary,
      completed_at: outcome.completed_at,
    })),
    cells: wire.slots.map((slot) => {
      const repair = repairBySlot.get(slot.evidence_slot_id);
      const gapCode = slot.typed_gap_codes[0];
      return {
        cell_id: slot.cell_id,
        decision_question: slot.decision_question,
        owner: slot.owner,
        materiality: slot.required ? "required" : "optional",
        evidence_slot_id: slot.evidence_slot_id,
        evidence_role: slot.evidence_role,
        request_id: slot.request_id,
        bundle_status: slot.bundle_status,
        candidates: slot.candidates.map((candidate) => ({
          ...candidate,
          evidence_slot_id: slot.evidence_slot_id,
          state: candidate.display_state,
          rejection_reason: candidate.review_reason,
        })),
        typed_gap: gapCode ? {
          gap_id: `${slot.bundle_id}:${gapCode}`,
          evidence_slot_id: slot.evidence_slot_id,
          state: repair ? "repair_requested" : "typed_gap",
          gap_code: gapCode,
          title: "No usable candidate",
          detail: "The bounded fixture route returned no candidate for this required evidence role.",
          stop_rule: "fixture_candidate_absent",
          repair_reason: repair?.reason,
          requested_at: repair?.recorded_at,
        } : null,
      };
    }),
  };
}

function evidencePath(caseId: string): string {
  return `${CASES_PATH}/${encodeURIComponent(caseId)}/evidence`;
}

function mutationInit(command: { idempotency_key: string }): RequestInit {
  return {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Idempotency-Key": command.idempotency_key,
    },
    body: JSON.stringify(command),
  };
}

function parsePayload(text: string): unknown {
  try {
    return JSON.parse(text) as unknown;
  } catch {
    return { detail: text };
  }
}

function toEvidenceApiError(statusCode: number, payload: unknown): EvidenceApiError {
  const envelope = errorEnvelope(payload);
  const nested = errorEnvelope(envelope.detail);
  const error = envelope.error ?? nested.error;
  return new EvidenceApiError({
    code: error?.error_code ?? envelope.code ?? nested.code ?? `http_${statusCode}`,
    message: error?.message ?? envelope.message ?? nested.message ?? detailMessage(envelope.detail) ?? "Evidence request failed.",
    statusCode,
    traceId: error?.trace_id ?? envelope.correlation_id ?? nested.correlation_id,
    detail: envelope.detail,
  });
}

function detailMessage(detail: unknown): string | undefined {
  if (typeof detail === "string") return detail;
  if (typeof detail !== "object" || detail === null) return undefined;
  const reason = (detail as { reason?: unknown }).reason;
  return typeof reason === "string" ? reason : undefined;
}

function errorEnvelope(value: unknown): ErrorEnvelope {
  return typeof value === "object" && value !== null ? value as ErrorEnvelope : {};
}
